### ISSUE TITLE
How does the --update parameter work? [question]

### ISSUE THREAD

**Issue by fschoenm** [conan-io/conan#5333]
I have problems understanding how the ``--update`` flag works for ``conan install``. My idea was that it always checks the remotes for new versions of a package. But in reality, the exact opposite seems to happen.

I built a new version of a package and uploaded it to my remote. As you can see, it's on both my local PC and the remote:

	$ conan upload -r sdk --all boost/*
	[...]

	$ conan search -r sdk boost
	Existing package recipes:

	boost/1.68.0@sdk/stable
	boost/1.69.0@sdk/stable
	boost/1.70.0@sdk/stable

	$ conan search boost
	Existing package recipes:

	boost/1.68.0@sdk/stable
	boost/1.69.0@sdk/stable
	boost/1.70.0@sdk/stable

But when calling ``conan install``, the behavior changes depending on if I use ``--update`` or not:

	$ conan install
	Configuration:
	[settings]
	arch=x86_64
	arch_build=x86_64
	build_type=Debug
	compiler=gcc
	compiler.libcxx=libstdc++11
	compiler.version=8
	os=Linux
	os_build=Linux
	[options]
	[build_requires]
	[env]
	CC=gcc-8
	CXX=g++-8
	Version ranges solved
	    Version range '>=1.68.0' required by 'conanfile.py' resolved to 'boost/1.70.0@sdk/stable'

	conanfile.py: Installing package
	Requirements
	    boost/1.70.0@sdk/stable from 'sdk' - Cache
	Packages
	    boost/1.70.0@sdk/stable:524ea35a8120baabdde02483add58d81bf541327 - Cache

	boost/1.70.0@sdk/stable: Already installed!
	boost/1.70.0@sdk/stable: LIBRARIES: []
	boost/1.70.0@sdk/stable: Package folder: /home/fschoenm/.conan/data/boost/1.70.0/sdk/stable/package/524ea35a8120baabdde02483add58d81bf541327
	conanfile.py: Generator cmake_multi created conanbuildinfo_debug.cmake
	conanfile.py: Generator cmake_multi created conanbuildinfo_multi.cmake
	conanfile.py: Generator txt created conanbuildinfo.txt
	conanfile.py: Generated conaninfo.txt
	conanfile.py: Generated graphinfo

	$ conan install .. --update
	Configuration:
	[settings]
	arch=x86_64
	arch_build=x86_64
	build_type=Debug
	compiler=gcc
	compiler.libcxx=libstdc++11
	compiler.version=8
	os=Linux
	os_build=Linux
	[options]
	[build_requires]
	[env]
	CC=gcc-8
	CXX=g++-8
	boost/1.69.0@sdk/stable: WARN: Setting 'cppstd' is deprecated in favor of 'compiler.cppstd', please update your recipe.
	Version ranges solved
	    Version range '>=1.68.0' required by 'conanfile.py' resolved to 'boost/1.69.0@sdk/stable'

	conanfile.py: Installing package
	Requirements
	    boost/1.69.0@sdk/stable from 'sdk' - Cache
	Packages
	    boost/1.69.0@sdk/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Cache

	boost/1.69.0@sdk/stable: Already installed!
	boost/1.69.0@sdk/stable: WARN: Lib folder doesn't exist, can't collect libraries: /home/fschoenm/.conan/data/boost/1.69.0/sdk/stable/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/lib
	boost/1.69.0@sdk/stable: LIBRARIES: []
	boost/1.69.0@sdk/stable: Package folder: /home/fschoenm/.conan/data/boost/1.69.0/sdk/stable/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9
	conanfile.py: Generator cmake_multi created conanbuildinfo_debug.cmake
	conanfile.py: Generator cmake_multi created conanbuildinfo_multi.cmake
	conanfile.py: Generator txt created conanbuildinfo.txt
	conanfile.py: Generated conaninfo.txt
	conanfile.py: Generated graphinfo

The conanfile.py requires boost >= 1.68.0:

	[...]
	self.requires("boost/[>=1.68.0]@sdk/stable")
	[...]

**My question:** Why does ``conan install --update`` resolve to an old version of boost while calling just ``conan install`` resolves to the new (correct) version? I don't really understand how the ``--update`` flag works. What do I have to do to correctly update a package, upload it to the remote, and have it automatically being used for all other developers using the same remote? What it does right now in my scenario above is the opposite of what I would assume to happen.

Using conan 1.16.0 on Ubuntu 19.04.

To help us debug your issue please explain:

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

**Comment by memsharded** [comment]
Hi @fschoenm 

I am running some tests, and so far I can't reproduce this behavior. This is my current attempt and it passes:
```python
    def update_remote_test(self):
        # https://github.com/conan-io/conan/issues/5333
        client = TestClient(servers={"default": TestServer()},
                            users={"default": [("lasote", "mypass")]})
        conanfile = textwrap.dedent("""
            from conans import ConanFile
            class Boost(ConanFile):
                pass
            """)
        client.save({"conanfile.py": conanfile})
        client.run("create . boost/1.68.0@lasote/stable")
        client.run("create . boost/1.69.0@lasote/stable")
        client.run("create . boost/1.70.0@lasote/stable")
        client.run("upload * -r=default --all --confirm")
        client.run("remove * -f")
        conanfile = textwrap.dedent("""
            [requires]
            boost/[>=1.68.0]@lasote/stable
            """)
        client.save({"conanfile.txt": conanfile}, clean_first=True)
        client.run("install .")
        self.assertIn("boost/1.70.0", client.out)
        self.assertNotIn("boost/1.69.0", client.out)
        self.assertNotIn("boost/1.68.0", client.out)

        client.run("install . --update")
        self.assertIn("boost/1.70.0", client.out)
        self.assertNotIn("boost/1.69.0", client.out)
        self.assertNotIn("boost/1.68.0", client.out)
```

There is a small different in your commands, in one case you are using ``conan install`` which refers to the conanfile in the current directory, and in the other you are doing ``conan install .. --update`` which refers to a conanfile in the parent directory. Is it possible that there is another conanfile there? In both cases, it is resolving it from the cache, not from the remote, not sure if this could be related. Any further idea that would help to reproduce? Thanks!

**Comment by memsharded** [comment]
Submitted the test to CI, just to make sure there is no difference in different platforms: https://github.com/conan-io/conan/pull/5336

**Comment by fschoenm** [comment]
@memsharded That was a type, I used ``conan install ..`` and there is no other conanfile than the one in the parent directory.

**Comment by fschoenm** [comment]
I found the issue: The problem was that I had an old remote in my configuration that didn't have the current version of the boost package. 

Unfortunately, I found that error only by using the conan trace file. Is there any way to make it more obvious in the output (or force in the input) where packages are coming from?

1. In the default output, there was no hint at all that conan was looking for updates to the packages on the wrong remote. Maybe conan could print out that and where it is looking for package updates if the parameter ``--update`` is specified?
2. How does conan prioritize multiple remotes? In my case, one of the remotes had the newer package but conan decided after querying one remote that there is no update.

**Comment by memsharded** [comment]
Hi @fschoenm 

The output shows the remote of each package. In the above test:

```conanfile.txt: Installing package
Requirements
    boost/1.70.0@lasote/stable from 'default' - Downloaded
Packages
    boost/1.70.0@lasote/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download

```

Where the ``default`` is the name of the remote repo.

> 2. How does conan prioritize multiple remotes? In my case, one of the remotes had the newer package but conan decided after querying one remote that there is no update.

The order of evaluation is the order in which the remotes are listed (``conan remote list``). When a match is found in one remote, that is used.

I will have a look to see if the output can be improved for suggestion 1)

**Comment by memsharded** [comment]
Adding more complete output to https://github.com/conan-io/conan/pull/5336, regarding the resolution of ranges.

**Comment by fschoenm** [comment]
> The output shows the remote of each package. In the above test:
> 
> ```
> Requirements
>     boost/1.70.0@lasote/stable from 'default' - Downloaded
> Packages
>     boost/1.70.0@lasote/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
> ```
> 
> Where the `default` is the name of the remote repo.

I think, this only works if the package is not cached locally like in my case. I had the following scenario:

Remote _sdk-fallback_:
- boost/1.68.0@sdk/stable
- boost/1.69.0@sdk/stable

Remote _sdk_:
- boost/1.68.0@sdk/stable
- boost/1.69.0@sdk/stable
- boost/1.70.0@sdk/stable

Locally cached:
- boost/1.69.0@sdk/stable
- boost/1.70.0@sdk/stable

My output (see above) didn't say anything about any remotes. It checked the old _sdk-fallback_ remote and found an old version of the package. Then it stopped because the package was already cached locally.

In fact, the two outputs looked exactly the same (apart from the version number obviously) even though what was happening behind the scenes was totally different (but only visible in the trace file).

Thanks for the answer and I hope you find a way to improve the output.

**Comment by memsharded** [comment]
Yes, the PR I have submitted will output something like:

```
Version ranges solved
    boost/*@lasote/stable resolved in 'default' remote
    Version range '>=1.68.0' required by 'conanfile.txt' resolved to 'boost/1.70.0@lasote/stable'
```
or
```
Version ranges solved
    boost/*@lasote/stable resolved in cache
    Version range '>=1.68.0' required by 'conanfile.txt' resolved to 'boost/1.70.0@lasote/stable'
```

I think that will help

**Comment by lasote** [comment]
The fix will be released at 1.17. Thanks!
